### PR TITLE
keep newBuilder with deprecation in Iterable

### PR DESCRIPTION
**Pull request by MasseGuillaume**
Fixes https://github.com/scala/bug/issues/11000

### CODE REVIEW

**Comment by julienrf** [comment]
The number given in the commit message is not correct, it should be scala/bug#11000 instead of just #11000.

**Comment by MasseGuillaume** [comment]
okay sorry: https://github.com/scala/scala/blob/2.13.x/CONTRIBUTING.md#bug-fix

File: test/junit/scala/collection/SearchingTest.scala [hunk 1]
```diff
@@ -15,7 +15,7 @@ class SearchingTest {
     class TestSeq[A](list: List[A]) extends Seq[A] {
       var elementsAccessed = immutable.Set.empty[Int]
 
-      protected[this] def newBuilder = ??? // not needed for this test
+      override protected[this] def newBuilder = ??? // not needed for this test
```
Comment (julienrf) [inline test/junit/scala/collection/SearchingTest.scala]: We can just remove this line, actually.

File: test/junit/scala/collection/SearchingTest.scala [hunk 2]
```diff
@@ -33,7 +33,7 @@ class SearchingTest {
     class TestIndexedSeq[A](vec: Vector[A]) extends IndexedSeq[A] {
       var elementsAccessed = immutable.Set.empty[Int]
 
-      protected[this] def newBuilder = ??? // not needed for this test
+      override protected[this] def newBuilder = ??? // not needed for this test
```
Comment (julienrf) [inline test/junit/scala/collection/SearchingTest.scala]: Same here.

**Comment by MasseGuillaume** [comment]
This makes unrelated test crash:

```
[info] Test scala.reflect.internal.MirrorsTest.rootCompanionsAreConnected started
java.lang.AssertionError: unsafe symbol Any (child of package scala) in runtime reflection universe
	at scala.reflect.internal.Symbols$Symbol.<init>(Symbols.scala:224)
	at scala.reflect.internal.Symbols$TypeSymbol.<init>(Symbols.scala:3066)
	at scala.reflect.internal.Symbols$ClassSymbol.<init>(Symbols.scala:3257)
	at scala.reflect.internal.Symbols$StubClassSymbol.<init>(Symbols.scala:3532)
	at scala.reflect.internal.Symbols.newStubSymbol(Symbols.scala:208)
	at scala.reflect.internal.Symbols.newStubSymbol$(Symbols.scala:204)
	at scala.reflect.internal.SymbolTable.newStubSymbol(SymbolTable.scala:17)
	at scala.reflect.internal.Symbols$Symbol.newStubSymbol(Symbols.scala:528)
	at scala.reflect.internal.pickling.UnPickler$Scan.$anonfun$readSymbol$6(UnPickler.scala:266)
	at scala.reflect.internal.pickling.UnPickler$Scan.$anonfun$readSymbol$5(UnPickler.scala:253)
	at scala.reflect.internal.pickling.UnPickler$Scan.$anonfun$readSymbol$4(UnPickler.scala:253)
	at scala.reflect.internal.pickling.UnPickler$Scan.readExtSymbol$1(UnPickler.scala:251)
	at scala.reflect.internal.pickling.UnPickler$Scan.readSymbol(UnPickler.scala:274)
	at scala.reflect.internal.pickling.UnPickler$Scan.readSymbolRef(UnPickler.scala:639)
	at scala.reflect.internal.pickling.UnPickler$Scan.readType(UnPickler.scala:406)
	at scala.reflect.internal.pickling.UnPickler$Scan.$anonfun$readTypeRef$1(UnPickler.scala:648)
	at scala.reflect.internal.pickling.UnPickler$Scan.at(UnPickler.scala:183)
	at scala.reflect.internal.pickling.UnPickler$Scan.readTypeRef(UnPickler.scala:648)
	at scala.reflect.internal.pickling.UnPickler$Scan.$anonfun$readType$1(UnPickler.scala:368)
	at scala.reflect.internal.pickling.PickleBuffer.until(PickleBuffer.scala:154)
	at scala.reflect.internal.pickling.UnPickler$Scan.readTypes$1(UnPickler.scala:366)
	at scala.reflect.internal.pickling.UnPickler$Scan.readType(UnPickler.scala:408)
	at scala.reflect.internal.pickling.UnPickler$Scan.$anonfun$readTypeRef$1(UnPickler.scala:648)
	at scala.reflect.internal.pickling.UnPickler$Scan.at(UnPickler.scala:183)
	at scala.reflect.internal.pickling.UnPickler$Scan.readTypeRef(UnPickler.scala:648)
	at scala.reflect.internal.pickling.UnPickler$Scan.readType(UnPickler.scala:410)
	at scala.reflect.internal.pickling.UnPickler$Scan$LazyTypeRef.$anonfun$completeInternal$1(UnPickler.scala:716)
	at scala.reflect.internal.pickling.UnPickler$Scan.at(UnPickler.scala:183)
	at scala.reflect.internal.pickling.UnPickler$Scan$LazyTypeRef.completeInternal(UnPickler.scala:716)
	at scala.reflect.internal.pickling.UnPickler$Scan$LazyTypeRef.complete(UnPickler.scala:743)
	at scala.reflect.internal.Symbols$Symbol.info(Symbols.scala:1527)
	at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$$anon$7.scala$reflect$runtime$SynchronizedSymbols$SynchronizedSymbol$$super$info(SynchronizedSymbols.scala:190)
	at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol.$anonfun$info$1(SynchronizedSymbols.scala:146)
	at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol.info(SynchronizedSymbols.scala:137)
	at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol.info$(SynchronizedSymbols.scala:146)
	at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$$anon$7.info(SynchronizedSymbols.scala:190)
	at scala.reflect.internal.pickling.UnPickler$Scan.fromName$1(UnPickler.scala:221)
	at scala.reflect.internal.pickling.UnPickler$Scan.$anonfun$readSymbol$3(UnPickler.scala:247)
	at scala.reflect.internal.pickling.UnPickler$Scan.readExtSymbol$1(UnPickler.scala:247)
	at scala.reflect.internal.pickling.UnPickler$Scan.readSymbol(UnPickler.scala:274)
	at scala.reflect.internal.pickling.UnPickler$Scan.readSymbolRef(UnPickler.scala:639)
	at scala.reflect.internal.pickling.UnPickler$Scan.readType(UnPickler.scala:406)
	at scala.reflect.internal.pickling.UnPickler$Scan.$anonfun$readTypeRef$1(UnPickler.scala:648)
	at scala.reflect.internal.pickling.UnPickler$Scan.at(UnPickler.scala:183)
	at scala.reflect.internal.pickling.UnPickler$Scan.readTypeRef(UnPickler.scala:648)
	at scala.reflect.internal.pickling.UnPickler$Scan.$anonfun$readType$1(UnPickler.scala:368)
	at scala.reflect.internal.pickling.PickleBuffer.until(PickleBuffer.scala:154)
	at scala.reflect.internal.pickling.PickleBuffer.until(PickleBuffer.scala:154)
	at scala.reflect.internal.pickling.UnPickler$Scan.readTypes$1(UnPickler.scala:366)
	at scala.reflect.internal.pickling.UnPickler$Scan.readType(UnPickler.scala:406)
	at scala.reflect.internal.pickling.UnPickler$Scan.$anonfun$readTypeRef$1(UnPickler.scala:648)
	at scala.reflect.internal.pickling.UnPickler$Scan.at(UnPickler.scala:183)
	at scala.reflect.internal.pickling.UnPickler$Scan.readTypeRef(UnPickler.scala:648)
	at scala.reflect.internal.pickling.UnPickler$Scan.readType(UnPickler.scala:410)
	at scala.reflect.internal.pickling.UnPickler$Scan$LazyTypeRef.$anonfun$completeInternal$1(UnPickler.scala:716)
	at scala.reflect.internal.pickling.UnPickler$Scan.at(UnPickler.scala:183)
	at scala.reflect.internal.pickling.UnPickler$Scan$LazyTypeRef.completeInternal(UnPickler.scala:716)
	at scala.reflect.internal.pickling.UnPickler$Scan$LazyTypeRef.complete(UnPickler.scala:743)
	at scala.reflect.internal.Symbols$Symbol.info(Symbols.scala:1527)
	at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$$anon$12.scala$reflect$runtime$SynchronizedSymbols$SynchronizedSymbol$$super$info(SynchronizedSymbols.scala:205)
	at 
```

**Comment by SethTisue** [comment]
the validation jobs on this PR (https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/3040/, https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/3041) produced gigabytes-long logfiles that gobbled up all the free space on our jenkins

**Comment by MasseGuillaume** [comment]
ops.

**Comment by SethTisue** [comment]
I had to delete those jobs to free up space, so the links are 404 now. but I grabbed one of the logfiles first, I'll take a look and post an excerpt here

**Comment by SethTisue** [comment]
oh, it's the same as the stack trace you already posted. I'll assume you're able to reproduce this locally, but if you need access to the full logfile, let me know.

**Comment by lrytz** [comment]
The failure looks like it could be related to / caused by https://github.com/scala/bug/issues/10903 (unfortunately...)

File: src/library/scala/collection/Iterable.scala [hunk 3]
```diff
@@ -26,6 +26,10 @@ trait Iterable[+A] extends IterableOnce[A] with IterableOps[A, Iterable, Iterabl
   protected def fromSpecificIterable(coll: Iterable[A @uncheckedVariance]): IterableCC[A] @uncheckedVariance = iterableFactory.from(coll)
   protected def newSpecificBuilder: Builder[A, IterableCC[A]] @uncheckedVariance = iterableFactory.newBuilder[A]
 
+  @deprecated("Use newSpecificBuilder instead", "2.13.0")
+  @deprecatedOverriding("Override newSpecificBuilder instead", "2.13.0")
```
Comment (sjrd) [inline src/library/scala/collection/Iterable.scala:30]: Will overriding `newBuilder` be sufficient to preserve the old behavior? If not, this `deprecatedOverriding` is not strong enough, and `newBuilder` must be `final` instead.

**Review by szeiger — CHANGES_REQUESTED** [review]
Making both `newSpecificBuilder` and `newBuilder` non-final leads to incorrect behavior. In such cases where we delegate in order to keep an old method name only one alias can be non-final. Since `newSpecificBuilder` must be overridable in subclasses, `newBuilder` has to be final.

Is it still worth adding it with this restriction? The original ticket doesn't provide a use case. It would be limited to very specific situations where a collection implementation is almost source compatible and has to call but not implement `newBuilder`.

**Comment by szeiger** [comment]
Closing because there doesn't appear to be a useful way forward.

**Pull request closed without merging**